### PR TITLE
apiextensions: publish (only) structural OpenAPI schemas

### PR DESCRIPTION
**Pull request by sttts**
This switch OpenAPI publishing over to structural schemas:

1) only structural schemas published
2) the go-openapi schema generation and OpenAPI v2 conversion goes through structural schema in order to simplify the logic and to remove go-openapi dependencies.

This is based on https://github.com/kubernetes/kubernetes/pull/77333. I.e. only the last commits are relevant here.

```
Publishing of CustomResourceDefinition validation schemas as part of /openapi/v2 (alpha feature, behind the CustomResourcePublishOpenAPI feature gate) is restricted to structural schemas only. CRDs with schemas which are not structural have the NonStructuralSchema condition set with the violations.
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go [hunk 1]
```diff
@@ -288,31 +291,35 @@ func (b *builder) buildRoute(root, path, action, verb string, sample interface{}
...
 	if schema == nil {
-		schema = &spec.Schema{
+		ret = &spec.Schema{
 			SchemaProps: spec.SchemaProps{Type: []string{"object"}},
 		}
 		// no, we cannot add more properties here, not even TypeMeta/ObjectMeta because kubectl will complain about
 		// unknown fields for anything else.
 	} else {
-		schema.SetProperty("metadata", *spec.RefSchema(objectMetaSchemaRef).
+		if v2 {
```
Comment (soltysh) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go]: This is always true from what I'm seeing, why we need it like that right now? Or am I missing something?

**Review by soltysh — APPROVED** [review]
One question, rest lgtm.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go]: We will publish v3 at some point. This is the start for that.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/unfold.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+			}
+
+			skipAnyOf := isIntOrStringAnyOfPattern(s)
+			skipFirstAllOfAnyOf := isIntOrStringAllOfPattern(s)
+			if skipAnyOf || skipFirstAllOfAnyOf {
+				return false
+			}
+
+			if s.AnyOf == nil {
+				s.AnyOf = []NestedValueValidation{
```
Comment (roycaihw) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/unfold.go]: currently for publishing v2 schema, the unfolded validation is dropped later by v3 -> v2 conversion?

File: test/integration/master/crd_test.go [hunk 3]
```diff
@@ -160,6 +160,7 @@ func TestCRDOpenAPI(t *testing.T) {
 			},
 			Validation: &apiextensionsv1beta1.CustomResourceValidation{
 				OpenAPIV3Schema: &apiextensionsv1beta1.JSONSchemaProps{
+					Type: "object",
```
Comment (roycaihw) [inline test/integration/master/crd_test.go]: is the root type still optional? or is there any new validation?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder_test.go [hunk 4]
```diff
@@ -22,56 +22,317 @@ import (
...
 			`{"$ref":"#/definitions/io.k8s.bar.v1.Foo"}`,
+			true,
 		},
-		{"type",
-			`{"type":"object"}`,
-			`{"properties":{"apiVersion":{},"kind":{},"metadata":{}},"type":"object","x-kubernetes-group-version-kind":[{"group":"bar.k8s.io","kind":"Foo","version":"v1"}]}`,
+		{"with extensions as v3 schema",
+			`
+{
+  "type":"object",
```
Comment (roycaihw) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder_test.go]: nit: extra trailing white space (also in L64)

**Review by roycaihw — COMMENTED** [review]
looks pretty good. I left some questions

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/unfold.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/unfold.go]: yes, I keep them complete here, with v3 publishing in mind we might want to add in 1.16.

File: test/integration/master/crd_test.go (hunk 3 above)
Comment (sttts) [reply test/integration/master/crd_test.go]: it's needed now for structural schemas, yes. We require structural schemas for publishing.

**Pull request merged**